File: 2025/1/1.pl
#!/usr/bin/env swipl
% usage: swipl 1.pl FILENAME
% https://adventofcode.com/2025/day/1

:- module(aoc2025day1, []).
:- module(aoc2025day1).
:- initialization(main, main).
:- use_module(library(clpfd)).
:- use_module(library(dcg/basics)).
:- use_module(library(dcg/high_order)).

load(Filename, Input) :- once(phrase_from_file(load(Input), Filename)).

load(Result) --> sequence(line, Result).
line(N) --> "L", integer(N0), eol, { N is -N0 }.
line(N) --> "R", integer(N), eol.

process(N, (D0, R0), (D, R)) :-
    D is (D0 + N) mod 100,
    (   D == 0
    ->  R is R0 + 1
    ;   R = R0
    ).

part_one(Input, Result) :- foldl(process, Input, (50, 0), (_, Result)).

main([File]) :-
    load(File, Input),
    part_one(Input, ResultOne),
    writeln(ResultOne).

File: 2025/1/1.jl
#!/usr/bin/env julia
# usage: julia 1.jl [FILENAME]
# https://adventofcode.com/2025/day/1

module AOC2025Day1

load(file::AbstractString) = [parse(Int, replace(l, 'L' => '-', 'R' => "")) for l in eachline(file)]

function partone(turns)
    total = 0
    dial = 50
    for turn in turns
        dial = mod(dial + turn, 100)
        if iszero(dial)
            total += 1
        end
    end
    total
end

function parttwo(turns)
    total = 0
    dial = 50
    for turn in turns
        while turn > 0
            dial = mod(dial + 1, 100)
            turn -= 1
            total += Int(iszero(dial))
        end
        while turn < 0
            dial = mod(dial - 1, 100)
            turn += 1
            total += Int(iszero(dial))
        end
    end
    total
end

if abspath(PROGRAM_FILE) == @__FILE__
    input = load(get(ARGS, 1, joinpath(@__DIR__, "input.txt")))
    println(partone(input))
    println(parttwo(input))
elseif isinteractive()
    using REPL
    REPL.activate(AOC2025Day1)
end

end # module
